autogen/oai/mistral.py

```diff
@@ -0,0 +1,275 @@
+            "model"
+        ], "Please specify the 'model' in your config list entry to nominate the Mistral.ai model to use."
+
+        # Validate allowed Mistral.AI parameters
+        mistral_params["stream"] = validate_parameter("stream", bool, False, False, None, [False])
+        mistral_params["temperature"] = validate_parameter("temperature", (int, float), True, 0.7, None, None)
+        mistral_params["top_p"] = validate_parameter("top_p", (int, float), True, None, None, None)
+        mistral_params["max_tokens"] = validate_parameter("max_tokens", int, True, None, (0, None), None)
+        mistral_params["safe_prompt"] = validate_parameter("safe_prompt", bool, False, False, None, [True, False])
+        mistral_params["random_seed"] = validate_parameter("random_seed", int, True, None, False, False)
+
+        return mistral_params
+
+    def create(self, params: Dict[str, Any]) -> ChatCompletion:
+        """Create a completion for a given config.
+
+        Args:
+            params: The params for the completion.
+
+        Returns:
+            The completion.
+        """
+        if "tools" in params:
+            converted_functions = params["tools"]
+        else:
+            converted_functions = None
+
+        raw_contents = params["messages"]
+
+        # Parse parameters to Mistral.AI API's parameters
+        mistral_params = self.parse_params(params)
+
+        mistral_messages = []
+        for message in raw_contents:
+
+            # Mistral
+            if message["role"] == "assistant" and "tool_calls" in message and message["tool_calls"] is not None:
+
+                # Convert OAI ToolCall to Mistral ToolCall
+                openai_toolcalls = message["tool_calls"]
+                mistral_toolcalls = []
+                for toolcall in openai_toolcalls:
+                    mistral_toolcall = ToolCall(id=toolcall["id"], function=toolcall["function"])
+                    mistral_toolcalls.append(mistral_toolcall)
+                mistral_messages.append(
+                    ChatMessage(role=message["role"], content=message["content"], tool_calls=mistral_toolcalls)
+                )
+
+            elif message["role"] in ("system", "user", "assistant", "tool"):
+                # Note this ChatMessage can take a 'name' but it is rejected by the Mistral API, so, no, the 'name' field is not used.
+                mistral_messages.append(ChatMessage(role=message["role"], content=message["content"]))
+
+            else:
+                warnings.warn(f"Unknown message role {message['role']}", UserWarning)
+
+        client = MistralClient(api_key=self.api_key)
+
+        # If a 'system' message follows an 'assistant' message, change it to 'user'
+        # This can occur when using LLM summarisation
+        for i in range(1, len(mistral_messages)):
+            if mistral_messages[i - 1].role == "assistant" and mistral_messages[i].role == "system":
+                mistral_messages[i].role = "user"
```

Comment:
If not changing roles, does it result in error?

Reply:
sometimes (it depends on the model templating).  I've been merging contents into the previous message if so, to avoid 2+ messages from same user.

Reply:
Through the API, yes, it will fail if we don't change roles.